src/pymor/algorithms/svd_va.py

```diff
@@ -0,0 +1,212 @@
+        returned.
+    rtol
+        Singular values smaller than this value multiplied by the largest singular
+        value are ignored.
+    atol
+        Singular values smaller than this value are ignored.
+    l2_err
+        Do not return more modes than needed to bound the l2-approximation
+        error by this value. I.e. the number of returned modes is at most ::
+
+            argmin_N { sum_{n=N+1}^{infty} s_n^2 <= l2_err^2 }
+
+        where `s_n` denotes the n-th singular value.
+    orthonormalize
+        If `True`, orthonormalize the computed POD modes again using
+        the :func:`~pymor.algorithms.gram_schmidt.gram_schmidt` algorithm.
+        If `False`, do not orthonormalize.
+    check
+        If `True`, check the computed POD modes for orthonormality.
+    check_tol
+        Tolerance for the orthonormality check.
+
+    Returns
+    -------
+    U
+        |VectorArray| of left singular vectors.
+    s
+        Sequence of singular values.
+    Vh
+        |NumPy array| of right singular vectors.
+    """
+    assert isinstance(A, VectorArrayInterface)
+    assert product is None or isinstance(product, OperatorInterface)
+    assert modes is None or modes <= len(A)
+
+    if A.dim == 0 or len(A) == 0:
+        return A.space.empty(), np.array([]), np.zeros((0, len(A)))
+
+    logger = getLogger('pymor.algorithms.svd_va.method_of_snapshots')
+
+    with logger.block(f'Computing Gramian ({len(A)} vectors) ...'):
+        B = A.gramian(product)
+
+    with logger.block('Computing eigenvalue decomposition ...'):
+        eigvals = (None
+                   if modes is None or l2_err > 0.
+                   else (len(B) - modes, len(B) - 1))
+
+        evals, V = spla.eigh(B, overwrite_a=True, turbo=True, eigvals=eigvals)
+        evals = evals[::-1]
+        V = V.T[::-1, :]
+
+        tol = max(rtol ** 2 * evals[0], atol ** 2)
+        above_tol = np.where(evals >= tol)[0]
+        if len(above_tol) == 0:
+            return A.space.empty(), np.array([]), np.zeros((0, len(A)))
+        last_above_tol = above_tol[-1]
+
+        errs = np.concatenate((np.cumsum(evals[::-1])[::-1], [0.]))
+        below_err = np.where(errs <= l2_err**2)[0]
+        first_below_err = below_err[0]
+
+        selected_modes = min(first_below_err, last_above_tol + 1)
+        if modes is not None:
+            selected_modes = min(selected_modes, modes)
+
+        s = np.sqrt(evals[:selected_modes])
+        V = V[:selected_modes]
+        Vh = V.conj()
+
+    with logger.block(f'Computing left-singular vectors ({len(V)} vectors) ...'):
+        U = A.lincomb(V / s[:, np.newaxis])
+
+    if orthonormalize:
+        with logger.block('Re-orthonormalizing left singular vectors ...'):
+            gram_schmidt(U, product=product, copy=False, check=False)
+        if len(U) < len(s):
+            raise AccuracyError('additional orthonormalization removed basis vectors')
```

Comment:
Does it really make sense (for an SVD algorithm) to re-othonormalize only the left-singular vectors? What about the right-singular vectors? Can the accuracy of the decomposition be negatively affected by only re-orthonormalizing the left-singular vectors?

Reply:
The right singular vectors come from `eigh`, which should be orthonormal.

Reply:
Ok, but is it clear that the accuracy of the overall decomposition does not suffer when re-orthonormalizing the left-singular vectors?

Reply:
In [Golub and Van Loan, Matrix Computations, fourth edition], section 8.6.3 begins with what would be the method of snapshots. The final step there is a QR decomposition with column pivoting of AV<sub>1</sub>, where V<sub>1</sub> is a matrix of eigenvectors of A<sup>T</sup>A. I didn't find the error analysis for this method in that book, and I don't know if there is any literature with numerical error analysis for the method of snapshots...